matter: samples: lock: new switchable variant (#13331)
nrfconnect/sdk-nrf: samples/matter/lock/README.rst

Thread and Wi-Fi switching
==========================

When built using the ``thread_wifi_switched`` build type and programmed to the nRF5340 DK with the nRF7002 EK shield attached, the door lock sample supports a feature that allows you to :ref:`switch between Matter over Thread and Matter over Wi-Fi <ug_matter_overview_architecture_integration_designs_switchable>` at runtime.
Due to Matter protocol limitations, a single Matter node can only use one transport protocol at a time.

.. matter_door_lock_sample_thread_wifi_switch_desc_start

> The application is built with support for both Matter and Thread and Matter over Wi-Fi.

Reviewer: ```suggestion
The application is built with support for both Matter over Thread and Matter over Wi-Fi.
```

Author: Good catch!

> The application is built with support for both Matter over Thread and Matter over Wi-Fi.

Reviewer 2: Please add information which transport is set by default.
Add information about how the button 1 - factory reset may impact on the application (if the device need to do some action to change to the previous transport)

Author: @wiba-nordic Just added two lines in this section to address @doublemis1 requests.

The device activates either Thread or Wi-Fi transport protocol on boot, based on a flag stored in the non-volatile memory on the device.
You can trigger the switch from one transport protocol to another using the **Button 3** on the nRF5340 DK.

> During the switch, the flag stored in the non-volatile memory is toggled, and then the device is factory reset and rebooted.

Reviewer: ```suggestion
This toggles the flag stored in the non-volatile memory, and then the device is factory reset and rebooted.
```

> The factory reset and recommissioning to a Matter fabric allows the device to be provisioned with network credentials for the current transport protocol, and start operating in the just selected network.

Reviewer: ```suggestion
The factory reset and recommissioning to a Matter fabric allows the device to be provisioned with network credentials for the transport protocol that it was switched to, and to start operating in the selected network.
```

Thread and Wi-Fi switching
==========================

When built using the ``thread_wifi_switched`` build type and programmed to the nRF5340 DK with the nRF7002 EK shield attached, the door lock sample supports a feature that allows you to :ref:`switch between Matter over Thread and Matter over Wi-Fi <ug_matter_overview_architecture_integration_designs_switchable>` at runtime.
Due to Matter protocol limitations, a single Matter node can only use one transport protocol at a time.

.. matter_door_lock_sample_thread_wifi_switch_desc_start

The application is built with support for both Matter over Thread and Matter over Wi-Fi.
The device activates either Thread or Wi-Fi transport protocol on boot, based on a flag stored in the non-volatile memory on the device.
By default, Matter over Wi-Fi is activated.

You can trigger the switch from one transport protocol to the other using the **Button 3** on the nRF5340 DK.

> This toggles the flag stored in the non-volatile memory, and then the device is factory reset and rebooted.
> The factory reset and recommissioning to a Matter fabric allows the device to be provisioned with network credentials for the transport protocol that it was switched to, and to start operating in the selected network.
> 
> The factory reset does not switch the device back to the default transport protocol (Wi-Fi).

Reviewer: ```suggestion
This toggles the flag stored in the non-volatile memory, and then the device is factory reset and rebooted.
Because the flag is toggled, the factory reset does not switch the device back to the default transport protocol (Wi-Fi).
Instead, the factory reset and recommissioning to a Matter fabric allows the device to be provisioned with network credentials for the transport protocol that it was switched to, and to start operating in the selected network.
```
Reorganized this one a bit.

      * If pressed for more than three seconds, it starts the NFC tag emulation, enables Bluetooth LE advertising for the predefined period of time (15 minutes by default), and makes the device discoverable over Bluetooth LE.

Button 3:

>     * On the nRF5340 DK when using the ``thread_wifi_switched`` build type: If pressed for more than ten seconds, it switches the Matter transport protocol from Thread or Wi-Fi to another and factory resets the device.

Reviewer: ```suggestion
    * On the nRF5340 DK when using the ``thread_wifi_switched`` build type: If pressed for more than ten seconds, it switches the Matter transport protocol from Thread or Wi-Fi to the other and factory resets the device.
```
Since we are only switching between two options.

Author: Ah, I saw two versions in the doc and wanted to unify them but I chose worse option :)

To test this feature, complete the following steps:

#. Build the door lock application using the ``thread_wifi_switched`` build type:

   .. code-block:: console

      west build -b nrf5340dk_nrf5340_cpuapp -- -DCONF_FILE=prj_thread_wifi_switched.conf -DSHIELD=nrf7002ek -Dmultiprotocol_rpmsg_SHIELD=nrf7002ek_coex

#. |connect_terminal_ANSI|
#. Program the application to the kit using the following command:

   .. code-block:: console

      west flash --erase

#. Wait until the device boots.
#. Find the following log message which indicates that Wi-Fi transport protocol is activated:

   .. code-block:: console

      D: 423 [DL]Starting Matter over Wi-Fi

> #. Press **Button 3** for more than ten seconds to trigger switching from one transport protocol to another.

Reviewer: ```suggestion
#. Press **Button 3** for more than ten seconds to trigger switching from one transport protocol to the other.
```

#. Wait until the operation is finished and the device boots again.

> #. Find the following log message which indicates that Wi-Fi transport protocol is activated:

Reviewer: ```suggestion
#. Find the following log message which indicates that Thread transport protocol is activated:
```
